[ML] Return 429 status when RequestExecutorService queue full

DonalEvans opened the pull request:
Rather than always throwing an ElasticsearchStatusException with 500 status in ActionUtils.wrapFailuresInElasticsearchException(), determine the appropriate status from the unwrapped exception

- Remove createInternalServerError() method from ActionUtils
- Refactor AlibabaCloudSearch*Action classes to be consistent with SenderExecutableAction
- Update tests to account for new behaviour

DonalEvans committed:
Return 429 status when RequestExecutorService queue full

Instead of returning a 500 (internal server error) response when the
RequestExecutorService queue is full and a new request is submitted,
return a 429 (too many requests) response.

- Wrap the existing EsRejectedExecutionException in an
  ElasticsearchStatusException before throwing
- Update existing tests for new behaviour

davidkyle commented:
Elasticsearch has code in the REST layer that catches any exception and generates a suitable status code for the exception 

https://github.com/elastic/elasticsearch/blob/4b7de2fa30f540e70dd6a69c9cc6adf1d19e220d/server/src/main/java/org/elasticsearch/rest/RestResponse.java#L122

Drilling into `ExceptionsHelper.status()` there is a check for `EsRejectedExecutionException` and it should return a 429

https://github.com/elastic/elasticsearch/blob/0350a48e41555bb7c714b35f7cb69f8dcdb6c878/server/src/main/java/org/elasticsearch/ExceptionsHelper.java#L71

Any `EsRejectedExecutionException` should be converted to a 429 status code, the question is why has that not happened. 

I hardcoded `RequestExecutorService` to throw a `EsRejectedExecutionException` and on `POST _inference/my_endpoint` I got the rejected error but with a status code 500, ie. has not been converted to a 429. The Elasticsearch REST API has an `error_trace` param that includes the exceptions stack trace (`POST _inference/my_endpoint?error_trace`) using that option I got this stack trace:

```
       "stack_trace": """org.elasticsearch.ElasticsearchStatusException: Failed to send X embeddings request. Cause: root cause rejected
	at org.elasticsearch.inference@9.2.0-SNAPSHOT/org.elasticsearch.xpack.inference.external.action.ActionUtils.createInternalServerError(ActionUtils.java:41)
	at org.elasticsearch.inference@9.2.0-SNAPSHOT/org.elasticsearch.xpack.inference.external.action.ActionUtils.lambda$wrapFailuresInElasticsearchException$0(ActionUtils.java:31)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:202)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:207)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:40)
	at org.elasticsearch.inference@9.2.0-SNAPSHOT/org.elasticsearch.xpack.inference.external.http.sender.TimedListener.lambda$getListener$1(TimedListener.java:49)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:266)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.ListenerTimeouts$TimeoutableListener.onFailure(ListenerTimeouts.java:96)
	at org.elasticsearch.inference@9.2.0-SNAPSHOT/org.elasticsearch.xpack.inference.external.http.sender.RequestTask.onRejection(RequestTask.java:59)
	at org.elasticsearch.inference@9.2.0-SNAPSHOT/org.elasticsearch.xpack.inference.external.http.sender.RequestExecutorService$RateLimitingEndpointHandler.enqueue(RequestExecutorService.java:495)
	at org.elasticsearch.inference@9.2.0-SNAPSHOT/org.elasticsearch.xpack.inference.external.http.sender.RequestExecutorService.execute(RequestExecutorService.java:342)
....
```

And that takes us to https://github.com/elastic/elasticsearch/blob/4dcc81f0c4acc0b047145292931abb7f3da9a85d/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/ActionUtils.java#L27

`EsRejectedExecutionException` is not an `ElasticsearchException` hence the 500 error code. 

`RequestExecutorService` is doing the right thing by returning a `EsRejectedExecutionException` but this translation logic in `ActionUtils` perhaps unnecessary. Please can you check if anything is relying on that code as it would be simplest to remove it and rely on the REST layer to generate the status code.

DonalEvans committed:
Revert "Return 429 status when RequestExecutorService queue full"

This reverts commit 05a2fe0ed3b6175a4a1dd03a98f544293b6d504e.

DonalEvans committed:
Determine appropriate status for wrapped exception

Rather than always throwing an ElasticsearchStatusException with 500
status in ActionUtils.wrapFailuresInElasticsearchException(), determine
the appropriate status from the unwrapped exception

- Remove createInternalServerError() method from ActionUtils
- Refactor AlibabaCloudSearch*Action classes to be consistent with
  SenderExecutableAction
- Update tests to account for new behaviour

DonalEvans committed:
Update docs/changelog/134178.yaml

DonalEvans committed:
Merge branch 'main' into return-429-when-queue-full

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/ActionUtils.java:
```diff
@@ -28,19 +27,17 @@ public static ActionListener<InferenceServiceResults> wrapFailuresInElasticsearc
+                    // Determine the appropriate RestStatus from the unwrapped exception, then wrap in an ElasticsearchStatusException
+                    new ElasticsearchStatusException(
+                        Strings.format("%s. Cause: %s", errorMessage, unwrappedException.getMessage()),
+                        ExceptionsHelper.status(unwrappedException),
```
👍 that's a neat solution

davidkyle commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/action/AlibabaCloudSearchCompletionAction.java:
```diff
@@ -61,16 +59,14 @@ public void execute(InferenceInputs inferenceInputs, TimeValue timeout, ActionLi
             return;
         }
 
+        ActionListener<InferenceServiceResults> wrappedListener = wrapFailuresInElasticsearchException(
```
These Alibaba classes are the outlier everything else is done by `SenderExecutableAction`. Please can you convert  `AlibabaCloudSearchActionCreator` to produce `SenderExecutableAction` and remove these classes. You will probably have to keep `AlibabaCloudSearchCompletionAction` as that has extra logic in it to check the inputs.

davidkyle reviewed (approved):
LGTM

I left a suggestion for some code cleanup with the Alibaba action classes. Please can you do that either as part of this PR or in another PR
